core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala

```diff
@@ -889,21 +891,40 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
     socket
   }
 
-  override def zero(value: JList[Array[Byte]]): JList[Array[Byte]] = new JArrayList
+  override def reset(): Unit = {
+    this._acc = new JArrayList[Array[Byte]]
+  }
+
+  override def isZero: Boolean = {
+    this._acc.isEmpty
+  }
 
-  override def addInPlace(val1: JList[Array[Byte]], val2: JList[Array[Byte]])
-      : JList[Array[Byte]] = synchronized {
+  override def copy(): PythonAccumulatorV2 = {
+    val newAcc = new PythonAccumulatorV2(serverHost, serverPort)
+    newAcc._acc = this._acc
+    newAcc
+  }
+
+  // This happens on the worker node, where we just want to remember all the updates
+  override def add(val2: JList[Array[Byte]]): Unit = synchronized {
+    _acc.addAll(val2)
+  }
+
+
+  override def merge(other: AccumulatorV2[JList[Array[Byte]], JList[Array[Byte]]]): Unit = {
+    val otherPythonAccumulator = other.asInstanceOf[PythonAccumulatorV2]
+    // This conditional isn't strictly speaking needed - merging only currently happens on the
+    // driver program - but that isn't gauranteed so incase this changes.
     if (serverHost == null) {
-      // This happens on the worker node, where we just want to remember all the updates
-      val1.addAll(val2)
-      val1
+      // We are on the worker
+      _acc.addAll(otherPythonAccumulator._acc)
     } else {
       // This happens on the master, where we pass the updates to Python through a socket
       val socket = openSocket()
       val in = socket.getInputStream
       val out = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream, bufferSize))
```

Comment:
On a bit of a side note - we could consider using the callback server here if we wanted to enable it in general rather than just for streaming once Py4J has its performance improvements in.